crates/throttle/src/throttle.rs

```diff
@@ -0,0 +1,366 @@
+        }
+    }
+
+    struct VanillaRedis(RedisConnection);
+
+    impl Throttler for VanillaRedis {
+        async fn throttle(
+            &self,
+            key: &str,
+            limit: u64,
+            period: Duration,
+            max_burst: u64,
+            quantity: Option<u64>,
+        ) -> Result<ThrottleResult, Error> {
+            redis_script_throttle(&self.0, key, limit, period, max_burst, quantity).await
+        }
+    }
+
+    async fn test_big_limits(
+        limit: u64,
+        max_burst: Option<u64>,
+        permitted_tolerance: f64,
+        throttler: &impl Throttler,
+    ) {
+        let period = Duration::from_secs(60);
+        let max_burst = max_burst.unwrap_or(limit);
+        let key = format!("test_big_limits-{limit}-{max_burst}");
+
+        let mut throttled_iter = None;
+
+        for i in 0..limit * 2 {
+            let result = throttler
+                .throttle(&key, limit, period, max_burst, None)
+                .await
+                .unwrap();
+            if result.throttled {
+                println!("iter: {i} -> {result:?}");
+                throttled_iter.replace(i);
+                break;
+            }
+        }
+
+        let throttled_iter = throttled_iter.expect("to hit the throttle limit");
+        let diff = ((max_burst as f64) - (throttled_iter as f64)).abs();
+        let tolerance = (max_burst as f64) * permitted_tolerance;
+        println!(
+            "throttled after {throttled_iter} iterations for \
+                 limit {limit}. diff={diff}. tolerance {tolerance}"
+        );
+        let max_rate = Rate::per_period(limit as i64, time::Duration::try_from(period).unwrap());
+        println!("max_rate: {max_rate:?}");
+
+        assert!(
+            diff < tolerance,
+            "throttled after {throttled_iter} iterations for \
+                limit {limit}. diff={diff} is not within tolerance {tolerance}"
+        );
+    }
+
+    #[tokio::test]
+    async fn basic_throttle_100() {
+        test_big_limits(100, None, 0.01, &*MEMORY).await;
+    }
+
+    #[tokio::test]
+    async fn basic_throttle_1_000() {
+        test_big_limits(1_000, None, 0.02, &*MEMORY).await;
+    }
+
+    #[tokio::test]
+    async fn basic_throttle_6_000() {
+        test_big_limits(6_000, None, 0.02, &*MEMORY).await;
+    }
+
+    #[tokio::test]
+    async fn basic_throttle_60_000() {
+        test_big_limits(60_000, None, 0.05, &*MEMORY).await;
+    }
+
+    #[tokio::test]
+    async fn basic_throttle_60_000_burst_30k() {
+        // Note that the 5% tolerance here is the same as the basic_throttle_60_000
+        // test case because the variance is due to timing issues with very small
+        // time periods produced by the overally limit, rather than the burst.
+        test_big_limits(60_000, Some(30_000), 0.05, &*MEMORY).await;
+    }
+
+    #[tokio::test]
+    async fn redis_throttle_1_000() {
```

Comment:
can we have one version of this that tries to use redis cell, that skips if redis cell is unavailable, and another that explicitly uses the script form, even if redis cell is available?